### ISSUE TITLE
[FEATURE] Unsloth ML Framework Integration — AI Agent-First Fine-Tuning Workflows on Colab GPUs

### ISSUE THREAD

**Issue by aryayt** [aryateja2106/LeCoder-cgpu-CLI#1]
## Feature Description

Integrate [Unsloth](https://github.com/unslothai/unsloth) — the state-of-the-art open-source LLM fine-tuning library — directly into the LeCoder cGPU CLI as a first-class feature. This enables AI agents and developers to **install, configure, and execute Unsloth-powered fine-tuning workflows entirely from the terminal**, without ever touching a browser or a notebook manually.

This also marks a strategic shift: **evolve LeCoder cGPU into a Colab-terminal-first platform** — not just a CLI wrapper, but a purpose-built agentic infrastructure layer for model training and fine-tuning via cloud GPUs.

---

## Problem Statement

Currently, fine-tuning LLMs with Unsloth requires:
1. Manually opening Google Colab in the browser
2. Cloning Unsloth notebooks, selecting GPU runtimes by hand
3. Running cells one-by-one with no programmatic control
4. No native way to trigger this from an AI agent, CI/CD, or terminal workflow

LeCoder cGPU already solves the GPU connectivity problem — but stops short of providing the **ML workflow layer** that agents and power users need to actually build, train, and fine-tune models. Unsloth is the perfect integration target because:
- It's 2x faster and uses 70% less VRAM than standard training
- It supports all major open-weight models (Llama, Qwen, Mistral, Phi, Gemma, etc.)
- It has Colab-native notebooks and HuggingFace integrations
- It's fully open source under Apache 2.0

---

### New Command Group: `lecoder-cgpu unsloth`

```bash
# Install Unsloth on the active Colab runtime
lecoder-cgpu unsloth install

# Verify Unsloth installation + GPU availability
lecoder-cgpu unsloth check

# Run a fine-tuning job using a preset notebook template
lecoder-cgpu unsloth finetune --model unsloth/Llama-3.2-3B-Instruct --dataset my_data.jsonl

# List available Unsloth model templates from HuggingFace
lecoder-cgpu unsloth models

# Run a specific Unsloth notebook from the official repo
lecoder-cgpu unsloth run --notebook Qwen3_(4B)-GRPO

# Download a fine-tuned adapter/model back to local
lecoder-cgpu unsloth download --output ./my-model

# Push fine-tuned model to HuggingFace Hub
lecoder-cgpu unsloth push --repo username/my-model-name
```

### New Command Group: `lecoder-cgpu finetune` (higher-level abstraction)

```bash
# One-command end-to-end fine-tune: connects to GPU, installs Unsloth, trains, and downloads
lecoder-cgpu finetune --model Llama-3.2-3B --dataset ./data.jsonl --output ./adapter

# AI agent-friendly JSON output for parsing
lecoder-cgpu finetune --model Qwen3-4B --json
```

---

## Key References & Sources

> **Important for implementing agents:** All links below are authoritative references. Do not hallucinate API details — read these sources directly.

### Unsloth
- 📦 **GitHub Repo**: https://github.com/unslothai/unsloth
- 📚 **Official Docs**: https://unsloth.ai/docs/get-started
- 🖥️ **VS Code + Colab Guide**: https://unsloth.ai/docs/get-started/install/vs-code
- 📓 **Notebooks Repo**: https://github.com/unslothai/notebooks
- 🤗 **HuggingFace Models**: https://huggingface.co/unsloth
- 🐳 **Docker Hub**: https://hub.docker.com/r/unsloth/unsloth
- 💬 **Reddit Community**: https://www.reddit.com/r/unsloth/

### Unsloth Install Commands (for Colab runtime)
```bash
# Standard Colab install
pip install unsloth

# With all extras for training
pip install "unsloth[colab-new] @ git+https://github.com/unslothai/unsloth.git"

# Verify install
python -c "import unsloth; print(unsloth.__version__)"
```

### Official Unsloth Notebook Templates (reference for `unsloth run` command)
- `Llama-3.2-3B` — Llama 3.2 3B Instruct fine-tuning
- `Qwen3_(4B)-GRPO` — Qwen3 4B Reinforcement Learning (GRPO)
- `Llama-3.1-8B` — Llama 3.1 8B instruction tuning
- `Mistral-7B` — Mistral 7B fine-tuning
- `Phi-3.5-mini` — Microsoft Phi-3.5 mini fine-tuning
- Full list: https://unsloth.ai/docs/get-started/unsloth-notebooks

### LeCoder cGPU CLI (this project)
- **Repo**: https://github.com/aryateja2106/LeCoder-cgpu-CLI
- **npm Package**: https://www.npmjs.com/package/lecoder-cgpu
- **Roadmap**: https://github.com/aryateja2106/LeCoder-cgpu-CLI/blob/main/ROADMAP.md
- **API Reference**: https://github.com/aryateja2106/LeCoder-cgpu-CLI/blob/main/docs/api-reference.md
- **AGENTS.md**: https://github.com/aryateja2106/LeCoder-cgpu-CLI/blob/main/AGENTS.md

### Related / Competing Projects to Study
- **Unsloth Studio** (official UI): https://unsloth.ai/studio
- **Google Colab VS Code Extension**: https://github.com/googlecolab/colab-vscode
- **Related GitHub Issue (VS Code kernel stuck bug)**: https://github.com/googlecolab/colab-vscode/issues/200

---

## Use Cases

1. **AI Agent Autonomy**: An agent receives a task like *"fine-tune Llama 3.2 on this JSONL dataset"* → calls `lecoder-cgpu finetune` → runs entirely without human browser interaction → downloads the adapter → pushes to HuggingFace.

2. **Student / Researcher Workflow**: A student with Colab Pro runs `lecoder-cgpu unsloth install && lecoder-cgpu unsloth run --notebook Qwen3-(4B)-GRPO` from their IDE terminal without touching the browser.

3. **CI/CD Model Training Pipeline**: A GitHub Actions workflow triggers `lecoder-cgpu finetune --model mistral-7b --dataset ./data.jsonl --json` and parses the output to track training metrics.

4. **Multi-Session Parallel Training**: Use the existing `--session` flag with the new `unsloth` commands to run fine-tuning on multiple Colab runtimes simultaneously (Colab Pro).

---

### Phase 1 — Unsloth Environment Bootstrap (v0.7.0 target)
- [ ] Add `lecoder-cgpu unsloth install` command that runs the Colab pip install via kernel mode
- [ ] Add `lecoder-cgpu unsloth check` to verify GPU compatibility and Unsloth version
- [ ] Store install state in execution history (`.lecoder-cgpu/unsloth-state.json`)
- [ ] Add `--unsloth` flag to existing `connect` and `run` commands to auto-install on session start

### Phase 2 — Notebook Template Execution (v0.7.0 target)
- [ ] Integrate with the Unsloth notebooks repo (`https://github.com/unslothai/notebooks`)
- [ ] Implement `lecoder-cgpu unsloth run --notebook <name>` to clone + execute an Unsloth notebook on Colab
- [ ] Add `lecoder-cgpu unsloth models` to list recommended models from `https://huggingface.co/unsloth`
- [ ] Extend `notebook create` templates with an `unsloth` template type

### Phase 3 — Fine-Tuning Abstraction Layer (v0.8.0 target)
- [ ] Implement `lecoder-cgpu finetune` as a high-level orchestration command
- [ ] Support SFTTrainer and GRPO training modes (Unsloth's two primary paradigms)
- [ ] Auto-detect dataset format (JSONL, CSV, HuggingFace dataset ID)
- [ ] Stream training loss/metrics to terminal in real-time
- [ ] Add `--push-to-hub` flag to upload fine-tuned model to HuggingFace Hub post-training
- [ ] `--json` output for agent parsing of training results

### Phase 4 — Agent-First Enhancements (v0.9.0 target)
- [ ] Emit structured events (token/loss/step) as JSON lines during training for agent consumption
- [ ] Add `lecoder-cgpu unsloth status` to poll training job progress
- [ ] Support `--background` flag to run training as a detached Colab job
- [ ] `lecoder-cgpu unsloth download --adapter` to pull LoRA adapters back to local disk
- [ ] HuggingFace Hub authentication integration (`huggingface-cli login` via kernel mode)

### Required Source Code Areas
- `src/commands/` — Add `unsloth.ts` and `finetune.ts` command modules
- `src/templates/` — Add Unsloth notebook templates
- `src/utils/` — Add `unsloth-installer.ts` and `hf-hub.ts` utilities
- `docs/` — Add `unsloth-guide.md` documentation
- `examples/` — Add fine-tuning workflow examples

---

## Alternatives Considered

- **Manual notebook execution**: Already possible via `lecoder-cgpu notebook open`, but not programmable enough for agents.
- **Local Unsloth install**: Requires expensive local GPU. Colab is free/affordable and already integrated.
- **Using `run` command manually**: Possible today (`lecoder-cgpu run "pip install unsloth"`), but no structured workflow or model management.

---

## Priority

- [x] **Critical** — blocks AI agent-driven model training workflows

---

## Additional Context

This feature directly aligns with the **v0.9.0 AI Integration** milestone in the roadmap (`ROADMAP.md`), which calls for:
> Enhanced AI agent features, Streaming output for LLMs, Conversation context preservation, Code generation helpers, Automatic error diagnosis

It also accelerates items in the **Future Considerations** section:
> ML Model Registry, Experiment Tracking (MLflow, W&B), Auto-scaling

The Unsloth VS Code + Colab integration (https://unsloth.ai/docs/get-started/install/vs-code) proves this workflow is already validated by Unsloth's own team — we are simply making it **fully programmatic and agent-accessible** from the terminal.

### Context for Implementing Agents

> If you are an AI agent working on this issue, follow these steps:
> 1. Read `AGENTS.md` first: https://github.com/aryateja2106/LeCoder-cgpu-CLI/blob/main/AGENTS.md
> 2. Study the existing `src/commands/` structure before adding new commands
> 3. Use kernel mode (`--mode kernel`) for all Python/pip operations on Colab
> 4. All new commands must support `--json` flag for structured output
> 5. Run `npm run build && lecoder-cgpu unsloth check` to verify your implementation
> 6. Add tests in `tests/` and integration tests in `integration-tests/`
> 7. Update `README.md`, `CHANGELOG.md`, and `docs/` with new command documentation

---

## Related Issues

- Roadmap v0.9.0 AI Integration: https://github.com/aryateja2106/LeCoder-cgpu-CLI/blob/main/ROADMAP.md#v090---advanced-features-q1-2026
- Unsloth VS Code + Colab Issue (kernel disconnect bug to handle): https://github.com/googlecolab/colab-vscode/issues/200
